src/formatting_stack/formatters/clean_ns.clj

```diff
@@ -64,32 +64,45 @@
      (make-cleaner how-to-ns-opts refactor-nrepl-opts namespaces-that-should-never-cleaned libspec-whitelist filename)
      how-to-ns-opts)))
 
+(defmacro with-memoized-libspec-allowlist
+  "Uses refactor-nrepl's 'memoized libspec' facility for best performance,
+  while respecting formatting-stack's need of lazily requiring refactor-nrepl."
+  {:style/indent 0}
+  [& body]
+  `(do
+     (require 'refactor-nrepl.ns.libspec-allowlist)
+     ((resolve 'refactor-nrepl.ns.libspec-allowlist/with-memoized-libspec-allowlist*)
+      (fn []
+        ~@body))))
+
 (defn format!
   [this files]
-  (->> files
-       (process-in-parallel! (fn [filename]
-                               (when-let [ns-replacement (replaceable-ns-form this filename)]
-                                 (println "Cleaning unused imports:" filename)
-                                 (write-ns-replacement! filename ns-replacement)))))
+  (with-memoized-libspec-allowlist
```

Comment:
This macro invocation could be a `strategy` instead added to the stack but I prefered to leave something that would survive any possible refactorings (as we foresee that will happen wrt config, strategies and stuff) and clearly work no matter what.

Could be revisited later when f-s' new API becomes an actual thing